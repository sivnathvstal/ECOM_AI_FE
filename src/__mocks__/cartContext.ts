import { Cart, CartProduct } from '../types';

export const mockCartItems: CartProduct[] = [
  {
    productId: 1,
    quantity: 2,
    product: {
      id: 1,
      title: 'Test Product 1',
      price: 10.99,
      description: 'Test description 1',
      category: 'electronics',
      image: 'test1.jpg',
      rating: { rate: 4.5, count: 100 },
    },
  },
  {
    productId: 2,
    quantity: 1,
    product: {
      id: 2,
      title: 'Test Product 2',
      price: 20.99,
      description: 'Test description 2',
      category: 'clothing',
      image: 'test2.jpg',
      rating: { rate: 4.0, count: 80 },
    },
  },
];

export const mockCart: Cart = {
  id: 1,
  userId: 1,
  date: '2023-01-01',
  products: [
    { productId: 1, quantity: 2 },
    { productId: 2, quantity: 1 },
  ],
};

export const mockCartContext = {
  cart: mockCart,
  cartItems: mockCartItems,
  isLoading: false,
  error: null,
  addItem: jest.fn(),
  updateItem: jest.fn(),
  removeItem: jest.fn(),
  totalItems: 3, // 2 + 1
  totalPrice: 42.97, // 10.99*2 + 20.99*1
};

export const mockEmptyCartContext = {
  cart: {
    id: 0,
    userId: 1,
    date: '2023-01-01',
    products: [],
  },
  cartItems: [],
  isLoading: false,
  error: null,
  addItem: jest.fn(),
  updateItem: jest.fn(),
  removeItem: jest.fn(),
  totalItems: 0,
  totalPrice: 0,
};

export const mockLoadingCartContext = {
  ...mockCartContext,
  isLoading: true,
};

export const mockErrorCartContext = {
  ...mockCartContext,
  error: 'Failed to fetch cart',
};

const defaultExport = {
  mockCartItems,
  mockCart,
  mockCartContext,
  mockEmptyCartContext,
  mockLoadingCartContext,
  mockErrorCartContext,
};
export default defaultExport; 