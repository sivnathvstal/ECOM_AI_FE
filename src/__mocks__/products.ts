import { Product, PaginationParams, ProductsResponse } from '../types';

export const mockProducts: Product[] = [
  {
    id: 1,
    title: 'Test Product 1',
    price: 10.99,
    description: 'Test description 1',
    category: 'electronics',
    image: 'test1.jpg',
    rating: { rate: 4.5, count: 100 },
  },
  {
    id: 2,
    title: 'Test Product 2',
    price: 20.99,
    description: 'Test description 2',
    category: 'clothing',
    image: 'test2.jpg',
    rating: { rate: 4.0, count: 80 },
  },
  {
    id: 3,
    title: 'Test Product 3',
    price: 15.99,
    description: 'Test description 3',
    category: 'electronics',
    image: 'test3.jpg',
    rating: { rate: 3.5, count: 60 },
  },
  {
    id: 4,
    title: 'Test Product 4',
    price: 25.99,
    description: 'Test description 4',
    category: 'jewelry',
    image: 'test4.jpg',
    rating: { rate: 4.8, count: 120 },
  },
];

export const mockPagination: PaginationParams = {
  page: 1,
  limit: 8,
  total: 4,
};

export const mockProductsResponse: ProductsResponse = {
  products: mockProducts,
  pagination: mockPagination,
};

export const mockCategories: string[] = [
  'electronics',
  'clothing',
  'jewelry',
]; 