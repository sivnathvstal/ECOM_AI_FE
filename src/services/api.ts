import axios from 'axios';

const API_URL = process.env.API_URL || 'https://fakestoreapi.com';

const api = axios.create({
  baseURL: API_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

export default api; 