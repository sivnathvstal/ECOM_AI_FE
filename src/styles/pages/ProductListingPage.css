.product-listing-container {
  padding: 16px;
}

.product-listing-title {
  margin-bottom: 16px;
}

.filter-container {
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-wrap: wrap;
  margin-bottom: 16px;
}

.search-container-wrapper {
  flex: 1;
  margin-right: 16px;
}

.loader-container {
  display: flex;
  justify-content: center;
  padding: 32px;
}

.no-results {
  text-align: center;
  padding: 32px;
} 