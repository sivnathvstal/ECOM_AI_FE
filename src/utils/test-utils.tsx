import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider, createTheme } from '@material-ui/core';
import { QueryClient, QueryClientProvider } from 'react-query';
import { CartProvider } from '../context/CartContext';

// Create a theme for testing
const theme = createTheme({
  palette: {
    primary: {
      main: '#3f51b5',
    },
    secondary: {
      main: '#f50057',
    },
  },
});

// Create a client for React Query
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

// All providers wrapper
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider theme={theme}>
        <CartProvider>
          <BrowserRouter>{children}</BrowserRouter>
        </CartProvider>
      </ThemeProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// Re-export everything
export * from '@testing-library/react';

// Override render method
export { customRender as render }; 