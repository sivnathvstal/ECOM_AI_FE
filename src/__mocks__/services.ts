import { mockProducts, mockProductsResponse, mockCategories } from './products';
import { mockCart, mockCartItems } from './cartContext';

// Product service mocks
export const mockProductService = {
  getProducts: jest.fn().mockResolvedValue(mockProductsResponse),
  getProductById: jest.fn().mockImplementation((id: number) => {
    const product = mockProducts.find(p => p.id === id);
    if (product) {
      return Promise.resolve(product);
    }
    return Promise.reject(new Error('Product not found'));
  }),
  getCategories: jest.fn().mockResolvedValue(mockCategories),
  searchProducts: jest.fn().mockResolvedValue(mockProductsResponse),
};

// Cart service mocks
export const mockCartService = {
  getCart: jest.fn().mockResolvedValue(mockCart),
  addToCart: jest.fn().mockResolvedValue(mockCart),
  updateCartItem: jest.fn().mockResolvedValue(mockCart),
  removeFromCart: jest.fn().mockResolvedValue(mockCart),
  getCartWithProducts: jest.fn().mockResolvedValue(mockCartItems),
}; 