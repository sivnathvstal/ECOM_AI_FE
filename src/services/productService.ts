import api from './api';
import { Product, ProductsResponse } from '../types';

export const getProducts = async (
  page = 1,
  limit = 8,
  category?: string
): Promise<ProductsResponse> => {
  try {
    // First get all products to calculate total
    const allProductsResponse = await api.get<Product[]>('/products');
    const allProducts = allProductsResponse.data;
    
    // Filter by category if provided
    const filteredProducts = category 
      ? allProducts.filter((product) => product.category === category) 
      : allProducts;
    
    const total = filteredProducts.length;
    
    // Calculate pagination
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + limit;
    const paginatedProducts = filteredProducts.slice(startIndex, endIndex);
    
    return {
      products: paginatedProducts,
      pagination: {
        page,
        limit,
        total
      }
    };
  } catch (error) {
    console.error('Error fetching products:', error);
    throw error;
  }
};

export const getProductById = async (id: number): Promise<Product> => {
  try {
    const response = await api.get<Product>(`/products/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error fetching product with id ${id}:`, error);
    throw error;
  }
};

export const getCategories = async (): Promise<string[]> => {
  try {
    const response = await api.get<string[]>('/products/categories');
    return response.data;
  } catch (error) {
    console.error('Error fetching categories:', error);
    throw error;
  }
};

export const searchProducts = async (
  query: string,
  page = 1,
  limit = 8
): Promise<ProductsResponse> => {
  try {
    const allProductsResponse = await api.get<Product[]>('/products');
    const allProducts = allProductsResponse.data;
    
    // Filter products by search query
    const filteredProducts = allProducts.filter((product) => 
      product.title.toLowerCase().includes(query.toLowerCase()) ||
      product.description.toLowerCase().includes(query.toLowerCase())
    );
    
    const total = filteredProducts.length;
    
    // Calculate pagination
    const startIndex = (page - 1) * limit;
    const endIndex = startIndex + limit;
    const paginatedProducts = filteredProducts.slice(startIndex, endIndex);
    
    return {
      products: paginatedProducts,
      pagination: {
        page,
        limit,
        total
      }
    };
  } catch (error) {
    console.error('Error searching products:', error);
    throw error;
  }
}; 