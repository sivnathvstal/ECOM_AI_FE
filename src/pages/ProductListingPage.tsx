import React, { useState, useEffect } from 'react';
import {
  Container,
  Grid,
  Typography,
  Box,
  CircularProgress
} from '@material-ui/core';
import ProductCard from '../components/product/ProductCard';
import SearchBar from '../components/common/SearchBar';
import CategoryFilter from '../components/common/CategoryFilter';
import Pagination from '../components/common/Pagination';
import { getProducts, searchProducts } from '../services/productService';
import { Product, PaginationParams } from '../types';
import './ProductListingPage.css';

const ProductListingPage: React.FC = () => {
  const [products, setProducts] = useState<Product[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const [error, setError] = useState<string | null>(null);
  const [searchQuery, setSearchQuery] = useState<string>('');
  const [selectedCategory, setSelectedCategory] = useState<string>('');
  const [pagination, setPagination] = useState<PaginationParams>({
    page: 1,
    limit: 8,
    total: 0,
  });

  const fetchProducts = async () => {
    try {
      setLoading(true);
      let response;
      
      if (searchQuery) {
        response = await searchProducts(searchQuery, pagination.page, pagination.limit);
      } else {
        response = await getProducts(pagination.page, pagination.limit, selectedCategory);
      }
      
      setProducts(response.products);
      setPagination(response.pagination);
      setError(null);
    } catch (err) {
      setError('Failed to fetch products');
      console.error(err);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchProducts();
  }, [pagination.page, selectedCategory, searchQuery]);

  const handlePageChange = (page: number) => {
    setPagination((prev) => ({ ...prev, page }));
  };

  const handleCategoryChange = (category: string) => {
    setSelectedCategory(category);
    setPagination((prev) => ({ ...prev, page: 1 })); // Reset to first page on category change
  };

  const handleSearch = (query: string) => {
    setSearchQuery(query);
    setPagination((prev) => ({ ...prev, page: 1 })); // Reset to first page on search
  };

  return (
    <Container className="product-listing-container" data-testid="product-listing-page">
      <Typography variant="h4" component="h1" className="product-listing-title" data-testid="page-title">
        Products
      </Typography>
      
      <Box className="filter-container">
        <Box className="search-container-wrapper">
          <SearchBar onSearch={handleSearch} initialQuery={searchQuery} />
        </Box>
        <CategoryFilter 
          selectedCategory={selectedCategory} 
          onCategoryChange={handleCategoryChange} 
        />
      </Box>
      
      {loading ? (
        <Box className="loader-container" data-testid="loading-indicator">
          <CircularProgress />
        </Box>
      ) : error ? (
        <Typography color="error" data-testid="error-message">
          {error}
        </Typography>
      ) : products.length === 0 ? (
        <Typography className="no-results" data-testid="no-results">
          No products found
        </Typography>
      ) : (
        <>
          <Grid container spacing={3} data-testid="product-grid">
            {products.map((product) => (
              <Grid item xs={12} sm={6} md={4} lg={3} key={product.id}>
                <ProductCard product={product} />
              </Grid>
            ))}
          </Grid>
          
          <Pagination 
            pagination={pagination} 
            onPageChange={handlePageChange} 
          />
        </>
      )}
    </Container>
  );
};

export default ProductListingPage; 