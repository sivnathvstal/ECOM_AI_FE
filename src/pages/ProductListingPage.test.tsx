import React from 'react';
import { render, screen, fireEvent, waitFor, act, within } from '../utils/test-utils';
import ProductListingPage from './ProductListingPage';
import * as productService from '../services/productService';
import { mockProductsResponse, mockProducts } from '../__mocks__/products';

// Mock the productService
jest.mock('../services/productService', () => ({
  getProducts: jest.fn(),
  searchProducts: jest.fn(),
  getCategories: jest.fn(),
}));

describe('ProductListingPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (productService.getProducts as jest.Mock).mockResolvedValue(mockProductsResponse);
    (productService.searchProducts as jest.Mock).mockResolvedValue(mockProductsResponse);
    (productService.getCategories as jest.Mock).mockResolvedValue(['electronics', 'clothing']);
  });

  it('renders correctly', async () => {
    let container;
    await act(async () => {
      const rendered = render(<ProductListingPage />);
      container = rendered.container;
    });
    
    await waitFor(() => expect(productService.getProducts).toHaveBeenCalled());
    expect(container).toMatchSnapshot();
  });

  it('displays loading state initially', async () => {
    
      render(<ProductListingPage />);
    
    expect(screen.getByTestId('loading-indicator')).toBeInTheDocument();
  });

  it('fetches and displays products', async () => {
    await act(async () => {
      render(<ProductListingPage />);
    });
    
    // Wait for products to be fetched
    await waitFor(() => expect(productService.getProducts).toHaveBeenCalled());
    
    // Check if products are rendered
    await waitFor(() => {
      mockProducts.forEach(product => {
        expect(screen.getByTestId(`product-card-${product.id}`)).toBeInTheDocument();
      });
    });
  });

  it('displays error message when API call fails', async () => {
    const error = new Error('Failed to fetch products');
    (productService.getProducts as jest.Mock).mockRejectedValue(error);
    
    await act(async () => {
      render(<ProductListingPage />);
    });
    
    await waitFor(() => {
      expect(screen.getByTestId('error-message')).toBeInTheDocument();
      expect(screen.getByTestId('error-message')).toHaveTextContent('Failed to fetch products');
    });
  });

  it('displays no results message when no products are found', async () => {
    (productService.getProducts as jest.Mock).mockResolvedValue({
      products: [],
      pagination: { page: 1, limit: 8, total: 0 }
    });
    
    await act(async () => {
      render(<ProductListingPage />);
    });
    
    await waitFor(() => {
      expect(screen.getByTestId('no-results')).toBeInTheDocument();
      expect(screen.getByTestId('no-results')).toHaveTextContent('No products found');
    });
  });

  it('searches for products when search query is submitted', async () => {
    await act(async () => {
      render(<ProductListingPage />);
    });
    
    // Wait for initial products to load
    await waitFor(() => expect(productService.getProducts).toHaveBeenCalled());
    
    // Find search input and submit a search query
    const searchInput = screen.getByTestId('search-input');
    const searchForm = screen.getByTestId('search-form');
    
    await act(async () => {
      fireEvent.change(searchInput, { target: { value: 'test query' } });
    });
    
    await act(async () => {
      fireEvent.submit(searchForm);
    });
    
    // Check if searchProducts was called with the correct parameters
    await waitFor(() => {
      expect(productService.searchProducts).toHaveBeenCalledWith('test query', 1, 8);
    });
  });

  it('filters products by category when category is selected', async () => {
    await act(async () => {
      render(<ProductListingPage />);
    });
    
    // Wait for initial products to load
    await waitFor(() => expect(productService.getProducts).toHaveBeenCalled());
    
    // Find category select and open dropdown
    const categorySelect = screen.getByTestId('category-select');
    fireEvent.mouseDown(categorySelect.querySelector('[role="button"]') || categorySelect);
    // Wait for the listbox to appear
    const listbox = await screen.findByRole('listbox');
    // Click on the electronics option
    const electronicsOption = within(listbox).getByTestId('category-electronics');
    fireEvent.click(electronicsOption);
    // Check if getProducts was called with the correct parameters
    await waitFor(() => {
      expect(productService.getProducts).toHaveBeenCalledWith(1, 8, 'electronics');
    });
  });

  it('changes page when pagination controls are used', async () => {
    // Override pagination for this test to ensure there are 2 pages
    (productService.getProducts as jest.Mock).mockResolvedValue({
      products: mockProducts,
      pagination: { page: 1, limit: 2, total: 4 },
    });
    await act(async () => {
      render(<ProductListingPage />);
    });
    // Wait for initial products to load
    await waitFor(() => expect(productService.getProducts).toHaveBeenCalled());
    // Find and click the next page button
    const nextPageButton = screen.getByTestId('next-page');
    await act(async () => {
      fireEvent.click(nextPageButton);
    });
    // Check if getProducts was called with the correct page
    await waitFor(() => {
      expect(productService.getProducts).toHaveBeenCalledWith(2, 2, '');
    });
  });
  it('calls setError(null) on successful fetch', async () => {
    await act(async () => {
      render(<ProductListingPage />);
    });
    await waitFor(() => expect(productService.getProducts).toHaveBeenCalled());
    expect(screen.queryByTestId('error-message')).not.toBeInTheDocument();
  });

  it('handles error in fetchProducts', async () => {
    (productService.getProducts as jest.Mock).mockRejectedValue(new Error('Failed to fetch products'));
    await act(async () => {
      render(<ProductListingPage />);
    });
    await waitFor(() => expect(screen.getByTestId('error-message')).toBeInTheDocument());
    expect(screen.getByTestId('error-message')).toHaveTextContent('Failed to fetch products');
  });
}); 