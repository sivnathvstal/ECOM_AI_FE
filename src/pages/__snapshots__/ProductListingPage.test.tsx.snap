// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListingPage renders correctly 1`] = `
<div>
  <div
    class="MuiContainer-root product-listing-container MuiContainer-maxWidthLg"
    data-testid="product-listing-page"
  >
    <h1
      class="MuiTypography-root product-listing-title MuiTypography-h4"
      data-testid="page-title"
    >
      Products
    </h1>
    <div
      class="MuiBox-root MuiBox-root-1 filter-container"
    >
      <div
        class="MuiBox-root MuiBox-root-2 search-container-wrapper"
      >
        <form
          class="MuiPaper-root search-container MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="search-form"
        >
          <div
            class="MuiInputBase-root search-input"
          >
            <input
              aria-label="search products"
              class="MuiInputBase-input"
              data-testid="search-input"
              placeholder="Search products..."
              type="text"
              value=""
            />
          </div>
          <button
            aria-label="search"
            class="MuiButtonBase-root MuiIconButton-root search-button"
            data-testid="search-button"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
      <div
        class="MuiFormControl-root category-filter"
        data-testid="category-filter"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          id="category-select-label"
        >
          Category
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          data-testid="category-select"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="category-select-label category-select"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
            id="category-select"
            role="button"
            tabindex="0"
          >
            <span>
              â€‹
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
      data-testid="product-grid"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
      >
        <div
          class="MuiPaper-root MuiCard-root product-card MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="product-card-1"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiCardActionArea-root"
            href="/product/1"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiCardMedia-root product-media"
              data-testid="product-image-1"
              style="background-image: url(test1.jpg);"
              title="Test Product 1"
            />
            <div
              class="MuiCardContent-root product-content"
            >
              <h2
                class="MuiTypography-root product-title MuiTypography-h6 MuiTypography-gutterBottom"
                data-testid="product-title-1"
              >
                Test Product 1
              </h2>
              <p
                class="MuiTypography-root product-price MuiTypography-body1 MuiTypography-colorTextPrimary"
                data-testid="product-price-1"
              >
                $
                10.99
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <div
            class="MuiCardActions-root product-actions MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
              data-testid="view-details-1"
              href="/product/1"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                View Details
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
      >
        <div
          class="MuiPaper-root MuiCard-root product-card MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="product-card-2"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiCardActionArea-root"
            href="/product/2"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiCardMedia-root product-media"
              data-testid="product-image-2"
              style="background-image: url(test2.jpg);"
              title="Test Product 2"
            />
            <div
              class="MuiCardContent-root product-content"
            >
              <h2
                class="MuiTypography-root product-title MuiTypography-h6 MuiTypography-gutterBottom"
                data-testid="product-title-2"
              >
                Test Product 2
              </h2>
              <p
                class="MuiTypography-root product-price MuiTypography-body1 MuiTypography-colorTextPrimary"
                data-testid="product-price-2"
              >
                $
                20.99
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <div
            class="MuiCardActions-root product-actions MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
              data-testid="view-details-2"
              href="/product/2"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                View Details
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
      >
        <div
          class="MuiPaper-root MuiCard-root product-card MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="product-card-3"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiCardActionArea-root"
            href="/product/3"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiCardMedia-root product-media"
              data-testid="product-image-3"
              style="background-image: url(test3.jpg);"
              title="Test Product 3"
            />
            <div
              class="MuiCardContent-root product-content"
            >
              <h2
                class="MuiTypography-root product-title MuiTypography-h6 MuiTypography-gutterBottom"
                data-testid="product-title-3"
              >
                Test Product 3
              </h2>
              <p
                class="MuiTypography-root product-price MuiTypography-body1 MuiTypography-colorTextPrimary"
                data-testid="product-price-3"
              >
                $
                15.99
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <div
            class="MuiCardActions-root product-actions MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
              data-testid="view-details-3"
              href="/product/3"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                View Details
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
      >
        <div
          class="MuiPaper-root MuiCard-root product-card MuiPaper-elevation1 MuiPaper-rounded"
          data-testid="product-card-4"
        >
          <a
            aria-disabled="false"
            class="MuiButtonBase-root MuiCardActionArea-root"
            href="/product/4"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiCardMedia-root product-media"
              data-testid="product-image-4"
              style="background-image: url(test4.jpg);"
              title="Test Product 4"
            />
            <div
              class="MuiCardContent-root product-content"
            >
              <h2
                class="MuiTypography-root product-title MuiTypography-h6 MuiTypography-gutterBottom"
                data-testid="product-title-4"
              >
                Test Product 4
              </h2>
              <p
                class="MuiTypography-root product-price MuiTypography-body1 MuiTypography-colorTextPrimary"
                data-testid="product-price-4"
              >
                $
                25.99
              </p>
            </div>
            <span
              class="MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root"
            />
          </a>
          <div
            class="MuiCardActions-root product-actions MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
              data-testid="view-details-4"
              href="/product/4"
              role="button"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                View Details
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-4 pagination-container"
      data-testid="pagination"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
        data-testid="previous-page"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </span>
          Previous
        </span>
      </button>
      <div
        class="MuiBox-root MuiBox-root-5 pagination-info"
        data-testid="page-info"
      >
        Page 
        1
         of 
        1
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
        data-testid="next-page"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Next
          <span
            class="MuiButton-endIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
